### ISSUE TITLE
Проблемы с зависимостями от булевого модификатора

### ISSUE THREAD

**Issue by Guria** [bem/bem-forum-content-ru#182]
https://gist.github.com/ec7034b7a409c826e989
Объявил зависимость shouldDeps модификатора от другого булевого модификатора.
```js
({
  shouldDeps:[
    { mods: { visible: true}}
  ]
})
```
В итоге в собранный deps.js булевый модификатор попадает дважды.
До исходного модификатора:
```js
    {
        "block": "drawer",
        "mod": "visible"
    },
```
После исходного модификатора:
```js
    {
        "block": "drawer",
        "mod": "visible",
        "val": true
    }
```
При этом появление модификатора visible до исходного является нежелательным и неожиданным.
Сборщик ENB настроенный ```generator-bem-stub```

**Comment by tadatuta** [comment]
// cc @andrewblond

**Comment by Guria** [comment]
Зависимость от модификатора со строковым значением тоже добавляет в deps.js 2 записи:
До:
```js
    {
        "block": "drawer",
        "mod": "visible"
    },
    {
        "block": "drawer",
        "mod": "side",
        "val": "right"
    },
```
После:
```js
    {
        "block": "drawer",
        "mod": "visible",
        "val": "full"
    }
```
Однако модификатор ```_visible_full``` будет вставлен уже в правильном месте, т.е. после исходного модификатора.

**Comment by andrewblond** [comment]
@Guria, привет!

По историческим причинам в `deps.js` формате информация о модификаторе добавляется 2 раза, вначале как модификатор по `*`, а потом по значению. Нужно это для одного редкого кейса:

```
block_mod.js      // общее поведение для всех модификаторов, 
                  // т.е. этот файл эквивалентен по смыслу
                  // такому названию `block_mod_*`
block_mod_val1.js // доопределяет `block_mod`
block_mod_val2.js // доопределяет `block_mod`
```

В данном случае `block_mod` рассматривается не как булевый, а как общий.

Если я правильно понял твою проблему, то тебе просто нужно использовать `mustDeps` вместо `shouldDeps`, чтобы `visible` модификатор был явно раньше, чем `drawer`.

**Comment by Guria** [comment]
@andrewblond `visible` у меня свойства `css`, которые должны быть после. Я вопрос решил, использовав модификатор со строковым значением.

Проблема, как я понял, в том, что булевый модификатор и реализация общего поведения для модификатора с разными значениями имеют "одинаковое имя файла".

**Comment by apsavin** [comment]
Должно быть так в любом случае:

block
block_mod
block_mod_v1
block_mod_v2

А у @Guria, как я понял, вот так:

block_mod
block

@Guria это так?

**Comment by andrewblond** [comment]
> Проблема, как я понял, в том, что булевый модификатор и реализация общего поведения для модификатора с разными значениями имеют "одинаковое имя файла".

Да, всё так. Логика такая: пока есть только один файл `block_mod` — это булевый модификатор,  как только добавляется файлов со значениями `block_mod_val`, то сразу следует рассматривать `block_mod` как общий.

В ENB нет возможности определить порядок зависимостей не подключая их, поэтому да, лучше использовать явное значение для каждого из модификаторов.
